### PR TITLE
Replace references to blueapi messaging package 

### LINKED ISSUE #588: Extract publishing of documents etc. to message bus into a separate library
...via @coretl:

As a developer I would like to install bluesky locally and develop plans/devices with my own `RunEngine` in a python shell or similar so that I can debug plans effectively without having to connect to a remote server (blueapi). I also need to ensure that the plan provokes the correct response from the rest of the system, so I will need easy-to-import helper functions that allow me to link my `RunEngine` to a message bus in the same way blueapi's is linked.

## Possible Solution

Extract the `messaging` package to a separate library, similar to https://github.com/bluesky/bluesky-kafka and provide helper functions that both blueapi and a standalone user in a shell can easily use to authenticate and connect to the message bus.

### PR DESCRIPTION
**Pull request by callumforrester**
Replace references to blueapi messaging package with references to the bluesky stomp library

Fixes #588

### CODE REVIEW

File: .github/actions/install_requirements/action.yml [hunk 1]
```diff
@@ -26,7 +26,11 @@ runs:
         python-version: ${{ env.PYTHON_VERSION }}
 
     - name: Install packages
-      run: pip install ${{ inputs.pip-install }}
+      run: pip install $([ -f dev-requirements.txt ] && echo '-c dev-requirements.txt') -e .[dev]
+      shell: bash
+
+    - name: Install bluesky_stomp
+      run: pip install git+https://github.com/DiamondLightSource/bluesky-stomp.git@main
```
Comment (ZohebShaikh) [inline .github/actions/install_requirements/action.yml:33]: This is for ad-hoc CI testing

File: dev-requirements.txt [hunk 2]
```diff
@@ -18,6 +18,7 @@ black==24.8.0
 bluesky==1.13.0a4
 bluesky-kafka==0.10.0
 bluesky-live==0.0.8
+bluesky-stomp @ git+https://github.com/DiamondLightSource/bluesky-stomp@4dbdb6b144b4b03243a1784f2f53dc13cbbef30e
```
Comment (ZohebShaikh) [inline dev-requirements.txt:21]: This is ad-hoc

File: tests/client/test_client.py [hunk 3]
```diff
@@ -307,9 +307,11 @@ def test_run_task_sets_up_control(
 ):
     mock_rest.create_task.return_value = TaskResponse(task_id="foo")
     mock_rest.update_worker_task.return_value = TaskResponse(task_id="foo")
+    ctx = Mock()
+    ctx.correlation_id = "foo"
+    mock_events.subscribe_to_all_events = lambda on_event: on_event(COMPLETE_EVENT, ctx)
```
Comment (ZohebShaikh) [inline tests/client/test_client.py:312]: I had to make this change to ensure that the task completes and doesn't hang indefinitely. I previously tried setting a short timeout, which allowed the test to finish, but it introduced an unintended dependency on `test_valid_stomp_config_for_listener` test. This is not ideal, as tests should be independent of each other. It seems the issue is related to the task handling.

```
        == "Subscribing to all bluesky events from localhost:61613\nPress enter to "
        + 'exit{\n  "state": "IDLE",\n  "task_status": null,\n  "errors": [],\n  '
        + '"warnings": []\n}\n'
```

File: pyproject.toml [hunk 4]
```diff
@@ -91,7 +91,7 @@ filterwarnings = ["error", "ignore::DeprecationWarning"]
 # Doctest python code in docs, python code in src docstrings, test functions in tests
 testpaths = "docs src tests"
 markers = [
-    "handler: marks tests that interact with the global handler object in handler.py",
+    "stomp: marks tests that require the stomp"
```
Comment (stan-dot) [inline pyproject.toml:94]: `the stomp` is not optimal, `the stomp client` would fit better

File: dev-requirements.txt (hunk 2 above)
Comment (DiamondJoseph) [reply dev-requirements.txt:21]: Did you have to make any changes to the bluesky-stomp library, or can we make a 0.1.0 on that repository?

File: src/blueapi/cli/cli.py [hunk 5]
```diff
@@ -147,14 +147,20 @@ def listen_to_events(obj: dict) -> None:
     config: ApplicationConfig = obj["config"]
     if config.stomp is not None:
         event_bus_client = EventBusClient(
-            StompMessagingTemplate.autoconfigured(config.stomp)
+            MessagingTemplate.for_broker(
```
Comment (stan-dot) [inline src/blueapi/cli/cli.py:150]: please rename it from Template, it's a client

Comment (stan-dot) [reply src/blueapi/cli/cli.py:150]: even with `import as`

File: src/blueapi/client/event_bus.py [hunk 6]
```diff
@@ -28,11 +30,11 @@ def __exit__(self, exc_type, exc_value, exc_traceback) -> None:
 
     def subscribe_to_all_events(
         self,
-        on_event: Callable[[MessageContext, AnyEvent], None],
+        on_event: Callable[[AnyEvent, MessageContext], None],
     ) -> None:
         try:
             self.app.subscribe(
-                self.app.destinations.topic("public.worker.event"),
+                MessageQueue(name="public.worker.event"),
```
Comment (stan-dot) [inline src/blueapi/client/event_bus.py:37]: this looks so much better now :rocket:

File: src/blueapi/service/interface.py [hunk 7]
```diff
@@ -51,10 +52,14 @@ def worker() -> TaskWorker:
 def messaging_template() -> MessagingTemplate | None:
     stomp_config = config().stomp
     if stomp_config is not None:
-        template = StompMessagingTemplate.autoconfigured(stomp_config)
+        template = MessagingTemplate.for_broker(
```
Comment (stan-dot) [inline src/blueapi/service/interface.py:55]: as above

File: dev-requirements.txt (hunk 2 above)
Comment (ZohebShaikh) [reply dev-requirements.txt:21]: I think we can release the bluesky-stomp library ,I did not make any changes on it

File: pyproject.toml (hunk 4 above)
Comment (DiamondJoseph) [inline pyproject.toml:94]:
```suggestion
    "stomp: marks tests that require a message bus configured for stomp" 
```

File: pyproject.toml [hunk 8]
```diff
@@ -24,7 +24,7 @@ dependencies = [
     "aiohttp",
     "PyYAML",
     "click",
-    "fastapi[all]",
+    "fastapi>=0.112.0",
```
Comment (DiamondJoseph) [inline pyproject.toml:27]:
```suggestion
    "fastapi[all]>=0.112.0",
```

unless there was a refactor of the dependencies? I don't know what we required from [all]

File: tests/core/fake_device_module.py [hunk 9]
```diff
@@ -33,7 +33,7 @@ def _mock_with_name(name: str) -> MagicMock:
 
 
 def wrong_return_type() -> int:
-    return "0"  # type: ignore
+    return "str"  # type: ignore
```
Comment (stan-dot) [inline tests/core/fake_device_module.py:36]: can we get a comment for this? where would we use this?

File: tests/service/test_interface.py [hunk 10]
```diff
@@ -273,3 +274,11 @@ def test_get_task_by_id(context_mock: MagicMock):
 def test_stomp_config():
     interface.set_config(ApplicationConfig(stomp=StompConfig()))
     assert interface.messaging_template() is not None
+
+
+@pytest.mark.stomp
+def test_stomp_connection_failure_with_invalid_config(caplog):
+    caplog.at_level(logging.exception)
+    interface.set_config(ApplicationConfig(stomp=StompConfig(host="Not Exists")))
```
Comment (stan-dot) [inline tests/service/test_interface.py:282]: please do not do 3 nested inits in one line

**Review by stan-dot — DISMISSED** [review]
small readability and naming suggestions

File: src/blueapi/client/event_bus.py (hunk 6 above)
Comment (DiamondJoseph) [inline src/blueapi/client/event_bus.py:37]:
```suggestion
                MessageTopic(name="public.worker.event"),
```

These were Topics (multiple consumers) before, not Queues (single consumers)

File: src/blueapi/service/interface.py [hunk 11]
```diff
@@ -51,10 +52,14 @@ def worker() -> TaskWorker:
...
+            broker=Broker(
+                host=stomp_config.host, port=stomp_config.port, auth=stomp_config.auth
+            )
+        )
 
         task_worker = worker()
-        event_topic = template.destinations.topic("public.worker.event")
+        event_topic = MessageQueue(name="public.worker.event")
```
Comment (DiamondJoseph) [inline src/blueapi/service/interface.py:62]:
```suggestion
        event_topic = MessageTopic(name="public.worker.event")
```

**Review by DiamondJoseph — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 8 above)
Comment (ZohebShaikh) [reply pyproject.toml:27]: I think we should remove the fastapi[all] as it is causing problems on main as well...I have created a PR to address that issue https://github.com/DiamondLightSource/blueapi/pull/607

File: tests/core/fake_device_module.py (hunk 9 above)
Comment (DiamondJoseph) [reply tests/core/fake_device_module.py]: This is unrelated to this change and will go away with a rebase on main. It's a result of [this change](https://github.com/DiamondLightSource/blueapi/pull/598) that made the handling of type conversion must more capable (by using the Pydantic handling rather than rolling our own).

**Comment by ZohebShaikh** [comment]
@callumforrester  Can you have a look at this PR, If everything looks good we can create a release of bluesky-stomp

File: src/blueapi/service/interface.py [hunk 12]
```diff
@@ -63,8 +68,12 @@ def messaging_template() -> MessagingTemplate | None:
...
             }
         )
-        template.connect()
-        return template
+        try:
+            template.connect()
+            return template
+        except Exception as ex:
+            logging.exception(msg="Failed to connect to message bus", exc_info=ex)
+            return None
```
Comment (DiamondJoseph) [inline src/blueapi/service/interface.py:76]: This except block changes the behaviour when a message bus is configured, but the message bus configuration is not valid.

Previously this was an error state that resulted in the failed startup of the service, which means on Kubernetes it would fall over, hopefully the message bus would recover, then this service could try and connect again. Now it will happily not create a connection in this case, which means on Kubernetes we will have created a declarative configuration but are not able to deliver that valid configuration.  I think this should be reverted and allowed to throw the exception here.

File: src/blueapi/service/main.py [hunk 13]
```diff
@@ -312,6 +312,8 @@ def set_state(
         and new_state in _ALLOWED_TRANSITIONS[current_state]
     ):
         if new_state == WorkerState.PAUSED:
+            if state_change_request.defer is None:
+                state_change_request.defer = False
```
Comment (DiamondJoseph) [inline src/blueapi/service/main.py:316]: Or can make StateChangeRequest.defer `bool` not `bool | None`- it's only going to be `None` if it's explicitly set to be but I don't see the use case for doing so?

**Review by DiamondJoseph — CHANGES_REQUESTED** [review]

File: tests/service/test_interface.py (hunk 10 above)
Comment (ZohebShaikh) [reply tests/service/test_interface.py:282]: ruff prefers it on one line

File: tests/conftest.py [hunk 14]
```diff
@@ -6,27 +6,6 @@
...
-        help="skip stomp tests (e.g. because a server is unavailable)",
-    )
-
-
-def pytest_configure(config):
-    config.addinivalue_line("markers", "stomp: mark test as requiring stomp broker")
-
-
-def pytest_collection_modifyitems(config, items):
-    if config.getoption("--skip-stomp"):
```
Comment (callumforrester) [inline tests/conftest.py:23]: Now that we've removed this, we no longer need any special handling for STOMP in CI, can you remove the parts of the github actions that start up activemq and rabbitmq?

Comment (ZohebShaikh) [reply tests/conftest.py:23]: There are 2 tests that are used to connect to the activemq/rabbitmq...I could delete them or skip both I don't think are suitable ..

**Comment by callumforrester** [comment]
@ZohebShaikh Would you be able to write some tests that cover the affected area? See https://app.codecov.io/gh/DiamondLightSource/blueapi/pull/589?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=DiamondLightSource

File: .github/workflows/_test.yml [hunk 15]
```diff
@@ -24,12 +24,6 @@ jobs:
...
     runs-on: ${{ inputs.runs-on }}
 
-    services:
-      activemq:
-        image: rmohr/activemq:5.14.5-alpine
-        ports:
-          - 61613:61613
-
     steps:
       - name: Start RabbitMQ
```
Comment (callumforrester) [inline .github/workflows/_test.yml:28]: Can you also remove this? I'd like to be sure it isn't needed before merging

File: src/blueapi/service/interface.py [hunk 16]
```diff
@@ -48,13 +49,17 @@ def worker() -> TaskWorker:
 
 
 @lru_cache
-def messaging_template() -> MessagingTemplate | None:
+def messaging_template() -> MessagingTemplate:
```
Comment (DiamondJoseph) [inline src/blueapi/service/interface.py:52]:
```suggestion
def messaging_template() -> MessagingTemplate | None:
```

This still returns None if stomp_config is None?

File: docs/reference/openapi.yaml [hunk 17]
```diff
@@ -1,481 +1,289 @@
...
-          anyOf:
-          - type: string
-          - type: 'null'
-          description: The ID of the current task, None if the worker is idle
-          title: Task Id
-      required:
-      - task_id
-      title: WorkerTask
-      type: object
+openapi: 3.1.0
```
Comment (DiamondJoseph) [inline docs/reference/openapi.yaml:1]: These changes seem unrelated, if possible I'd like the openapi.yaml to remain unchanged, or at least reduce the size of the changes to this file. Do we know why it needs to be regenerated? If it isn't are there changes?

**Review by DiamondJoseph — CHANGES_REQUESTED** [review]

**Review by DiamondJoseph — APPROVED** [review]
LGTM!
